File: LeetCode/merge_K_Sorted_Lists.java

/**
 * You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.

Merge all the linked-lists into one sorted linked-list and return it.

 

Example 1:

Input: lists = [[1,4,5],[1,3,4],[2,6]]
Output: [1,1,2,3,4,4,5,6]
Explanation: The linked-lists are:
[
  1->4->5,
  1->3->4,
  2->6
]
merging them into one sorted list:
1->1->2->3->4->4->5->6
Example 2:

Input: lists = []
Output: []
Example 3:

Input: lists = [[]]
Output: []
 

Constraints:

k == lists.length
0 <= k <= 104
0 <= lists[i].length <= 500
-104 <= lists[i][j] <= 104
lists[i] is sorted in ascending order.
The sum of lists[i].length will not exceed 104.
 */

import java.util.*;

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class merge_K_Sorted_Lists {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> heap = new PriorityQueue<>((a,b)->a.val-b.val);
        for(ListNode i: lists){
            if(i!=null) heap.offer(i);
        }
        ListNode list = new ListNode();
        ListNode head = list;
        while(!heap.isEmpty()){
            ListNode temp = heap.poll();
            list.next = temp;
            temp = temp.next;
            list = list.next;
            list.next = null;
            if(temp!=null)
            heap.offer(temp);
        }

        return head.next;
    }
}

File: Leetcode-solution-python/mergeKSortedLists.py
# https://leetcode.com/problems/merge-k-sorted-lists/

from typing import Optional, List
from ListNode import ListNode
from heapq import heappush, heappop


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap, ans = [], ListNode(-1)
        for i, temp in enumerate(lists):
            if temp:
                heappush(heap, (temp.val, i, temp))
        new_head = ans
        while heap:
            temp = heappop(heap)
            ans.next = temp[2]
            if temp[2].next:
                heappush(heap, (temp[2].next.val, temp[1], temp[2].next))
            ans = ans.next
            ans.next = None

        return new_head.next
